AES CTR-DRGB: use platform specific AES-CTR support

p-steuer opened the pull request:
Next try to get HW supported AES-CTR for CTR-DRBG's generate operation.

Related:

- My original 2017 PR: https://github.com/openssl/openssl/pull/4530
(was rejected/stalled due to not using EVP and worries about internal state's counter increment not exactly matching the NIST pseudo-code/reference implementation)

- Today's PR: https://github.com/openssl/openssl/pull/10456
(tries to do the same only for x86)


This one uses EVP, keeps internal state counter in sync with reference implementation and should already pass the test suite.

TODO:
- simplyfiy if-statement in generate for loop - done.
- look at performance especially for small chunks - done.
- add test case(s) that cover counter wrapping (think my old PR had that one) - done.

p-steuer commented:
Not looking good ...
```
                 master         this PR
[bytes]         [bytes/sec]    [bytes/sec]
   16              6044.29k       4689.63k
   24              8809.23k       6837.68k
   32             11773.52k       8431.58k
   64             22041.49k      16890.84k
  256             63664.31k      68692.49k
  512             97853.70k     140839.77k
16384            184698.47k    2066346.26k

master (16 bytes)
  16.22%  openssl  libc-2.17.so        [.] __strcmp_sse42             
  13.17%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt          
  10.73%  openssl  libcrypto.so.3      [.] EVP_CIPHER_block_size      
   8.21%  openssl  libcrypto.so.3      [.] OSSL_PARAM_locate          
   5.00%  openssl  libcrypto.so.3      [.] cipher_generic_block_update
   3.97%  openssl  libcrypto.so.3      [.] cipher_generic_get_params  
   3.92%  openssl  libcrypto.so.3      [.] aesni_set_encrypt_key      
   3.44%  openssl  libcrypto.so.3      [.] EVP_EncryptUpdate          
   2.87%  openssl  libpthread-2.17.so  [.] pthread_rwlock_unlock      
   2.14%  openssl  libpthread-2.17.so  [.] pthread_rwlock_wrlock      
   2.07%  openssl  libcrypto.so.3      [.] OPENSSL_cleanse            
   1.91%  openssl  libc-2.17.so        [.] __memset_sse2              
   1.53%  openssl  libcrypto.so.3      [.] fillblock                  
   1.52%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_key_length  
   1.48%  openssl  libcrypto.so.3      [.] ctr_df                     
   1.45%  openssl  libc-2.17.so        [.] _int_free                  
   1.41%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_reset       
   1.32%  openssl  libcrypto.so.3      [.] strcmp@plt                 
   1.04%  openssl  libc-2.17.so        [.] _int_malloc                
   1.02%  openssl  libpthread-2.17.so  [.] pthread_rwlock_rdlock 

this PR (16 bytes)
  14.60%  openssl  libc-2.17.so        [.] __strcmp_sse42             
   9.59%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt          
   8.20%  openssl  libcrypto.so.3      [.] EVP_CIPHER_block_size      
   7.42%  openssl  libcrypto.so.3      [.] OSSL_PARAM_locate          
   5.97%  openssl  libcrypto.so.3      [.] aesni_set_encrypt_key      
   3.54%  openssl  libcrypto.so.3      [.] cipher_generic_block_update
   3.37%  openssl  libcrypto.so.3      [.] OPENSSL_cleanse            
   3.29%  openssl  libpthread-2.17.so  [.] pthread_rwlock_unlock      
   3.21%  openssl  libcrypto.so.3      [.] cipher_generic_get_params  
   2.89%  openssl  libc-2.17.so        [.] __memset_sse2              
   2.77%  openssl  libpthread-2.17.so  [.] pthread_rwlock_wrlock      
   2.63%  openssl  libcrypto.so.3      [.] EVP_EncryptUpdate          
   2.32%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_key_length  
   2.10%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_reset       
   1.92%  openssl  libc-2.17.so        [.] _int_free                  
   1.53%  openssl  libcrypto.so.3      [.] EVP_CipherInit_ex          
   1.52%  openssl  libc-2.17.so        [.] _int_malloc                
   1.27%  openssl  libc-2.17.so        [.] malloc                     
   1.22%  openssl  libcrypto.so.3      [.] strcmp@plt                 
   1.07%  openssl  libcrypto.so.3      [.] ctr_df                     
   1.07%  openssl  libcrypto.so.3      [.] fillblock                       

master (16384 bytes)
  21.53%  openssl  libc-2.17.so        [.] __strcmp_sse42             
  18.58%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt          
  16.59%  openssl  libcrypto.so.3      [.] EVP_CIPHER_block_size      
  10.78%  openssl  libcrypto.so.3      [.] OSSL_PARAM_locate          
   7.44%  openssl  libcrypto.so.3      [.] cipher_generic_block_update
   6.19%  openssl  libcrypto.so.3      [.] cipher_generic_get_params  
   5.23%  openssl  libcrypto.so.3      [.] EVP_EncryptUpdate          
   2.30%  openssl  libcrypto.so.3      [.] fillblock                  
   1.82%  openssl  libcrypto.so.3      [.] strcmp@plt                 
   1.47%  openssl  libcrypto.so.3      [.] drbg_ctr_generate          
   1.47%  openssl  libcrypto.so.3      [.] cipher_hw_aesni_ecb        
   1.14%  openssl  libcrypto.so.3      [.] trailingdata               
   1.07%  openssl  libcrypto.so.3      [.] OSSL_PARAM_set_uint64   

this PR (16384)
  51.91%  openssl  libcrypto.so.3      [.] aesni_ctr32_encrypt_blocks                                                                                                                          
   6.49%  openssl  libc-2.17.so        [.] __strcmp_sse42                                                                                                                                      
   5.39%  openssl  libc-2.17.so        [.] __memset_sse2                                                                                                                                       
   4.45%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt                                                                                                                                   
   3.54%  openssl  libcrypto.so.3      [.] EVP_CIPHER_block_size                                                                                                                               
   3.23%  openssl  libcrypto.so.3      [.] OSSL_PARAM_locate                                                                                                                                   
   2.56%  openssl  libcrypto.so.3      [.] aesni_set_encrypt_key                                                                                                                               
   1.65%  openssl  libcrypto.so.3      [.] cipher_generic_block_update                                                                                                                         
   1.42%  openssl  libcrypto.so.3      [.] OPENSSL_cleanse                                                                                                                                     
   1.41%  openssl  libpthread-2.17.so  [.] pthread_rwlock_unlock                                                                                                                               
   1.38%  openssl  libcrypto.so.3      [.] cipher_generic_get_params                                                                                                                           
   1.21%  openssl  libcrypto.so.3      [.] EVP_EncryptUpdate                                                                                                                                   
   1.19%  openssl  libpthread-2.17.so  [.] pthread_rwlock_wrlock      
```

The overhead of managing 2 EVP contexts instead of one (ECB and CTR) makes this PR slower than master for generate operations < ~250 bytes. Afterwards, there is a performance gain, but its several dimensions smaller that the old PR's which used the low level interfaces. The high amounts of strcmp catches the eye but even if that could be reduced its still a lot of overhead: aes HW has become so fast these days that even a deep callstack is payed with >1% per frame.

slontis commented:
Hopefully #10461 will get rid of the EVP_CIPHER_block_size() performance issue..

raja-ashok commented on crypto/rand/drbg_ctr.c:
```diff
@@ -287,7 +296,10 @@ __owur static int drbg_ctr_instantiate(RAND_DRBG *drbg,
+    if (!EVP_CipherInit_ex(ctr->ctx_ecb,
+                           ctr->cipher_ecb, NULL, ctr->K, NULL, 1)
+        || !EVP_CipherInit_ex(ctr->ctx_ctr,
+                              ctr->cipher_ctr, NULL, ctr->KX, NULL, 1))
```
I feel here `ctr->ctx_ctr` key initialization is not needed. Because anyway after this `ctr_update()` initializes `ctr->ctx_ctr` with updated `ctx->K`

raja-ashok commented on crypto/rand/drbg_ctr.c:
```diff
@@ -199,20 +200,23 @@ __owur static int ctr_df(RAND_DRBG_CTR *ctr,
+    if (!EVP_CipherInit_ex(ctr->ctx_ecb,
+                           ctr->cipher_ecb, NULL, ctr->KX, NULL, 1)
+        || !EVP_CipherInit_ex(ctr->ctx_ctr,
+                              ctr->cipher_ctr, NULL, ctr->KX, NULL, 1))
```
I feel here also initializing `ctx->ctr` is not needed. Only initializing at the end of `ctr_update` will be sufficient I feel.

p-steuer replied:
You are right in both cases. I pushed a fixup and updated the perf data in my previous post. It looks slightly better now.

p-steuer commented:
Updated the performance data with fixup included. Looks slightly better.

raja-ashok commented:
Combining multiple `EVP_EncryptUpdate` calls improves the rand generation performance by 30%. The code changes is available [here](https://github.com/raja-ashok/openssl/commit/76ea441fbda771bb3b86b30c122c1579e8983a5e).

## With `master`
```
type     16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
rand      4688.79k    17899.91k    56310.81k   119073.90k   172064.99k   178398.67k
```

## With #10457
```
type     16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
rand      4029.92k    16387.68k    62816.27k   243726.63k  1233314.13k  1677456.98k
```

## With #10457 and #10461
```
type     16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
rand      5343.63k    21191.28k    82521.14k   307550.51k  1489285.62k  2064954.62k
```

## With #10457, #10461 and [new changes](https://github.com/raja-ashok/openssl/commit/76ea441fbda771bb3b86b30c122c1579e8983a5e)
```
type     16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
rand      6936.96k    27967.18k   106227.63k   388237.87k  1711259.94k  2259579.59k
```

paulidale commented:
That looks like a decent speed up.

kroeckx commented:
But still seems like a factor 2 slower than the original PR.

p-steuer commented:
@raja-ashok  may i push your commit to this branch such that it is included in this PR ?

raja-ashok commented:
@p-steuer Yes please.

p-steuer commented:
pushed raja's optimization and a small strict-warnings fixup and updated my second post with performance numbers and sampling results (https://github.com/openssl/openssl/pull/10461 not included, i ll do that once its merged and this PR rebased).

Perf looks strictly better than current master. However, small chunks are still a lot slower than old master (for small chunks) and my old PR. Lets see how it looks after   https://github.com/openssl/openssl/pull/10461 is merged.

p-steuer commented:
Merged https://github.com/openssl/openssl/pull/ (was "ready to merge") and rebased this one.

Updated performance numbers and sampling data in my second post.

p-steuer commented:
@raja-ashok test_rand fails with your changes. Looking into it.

p-steuer commented:
Pushed a fixup. Test suite should be passed now. I had to undo the changes to ctr_df to make it work. Updated second post numbers.

kroeckx commented:
So is this ready to be reviewed, it's still marked as WIP, or do you intend to make other changes?

p-steuer commented:
I intend to add a test case for counter wrapping. Otherwise the code itself is ready to be reviewed. Removing the WIP .

mspncp commented:
> Good work with the speedup.
> I think this code could do with some general comments - to make it obvious that you are still following the steps required to do the algorithm for DRBG-CTR.

ping @p-steuer

p-steuer commented:
Simplyfied code a little, squashed everything and added a test case for 32-bit counter wrap.

@slontis , @mspncp  okay , i ll add some comments

 .. and i ll provide some final performance numbers.

p-steuer commented:
think travis fails (evp tests fail) are not due to this PR.

p-steuer commented:
Rebased and put an extensive description of the changes in the commit message (I chose this option over putting comments in the code because i found the present code base to already deviate from the standards exact pseudo code).

I ll provide some perf data tomorrow and this should be ready to go.

p-steuer commented:
```
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes

x86_64 (CPUINFO: OPENSSL_ia32cap=0x7ffaf3ffffebffff:0x29c6fbf)
rand pre -patch     10049.06k    37961.03k   125751.81k   296885.25k   485537.11k   518548.14k
rand post-patch      9492.44k    37744.90k   122006.19k   507212.80k  2046580.05k  2672421.55k

s390 (z15)
rand pre -patch         7577.56k    27961.56k    88870.05k   192584.85k   294301.02k   305119.23k
rand post-patch        8381.95k    33294.24k   131116.87k   514722.31k  3110873.25k  4649911.60k
```

For small sizes, the overhead (strcmp,malloc..) of managing 2 ctx seems to eat up the theoretic perf. gain, at least for x86 (samples from x86_64, 16 byte size):
```
pre patch
  17.56%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt                                                                                                                                   
   9.44%  openssl  libcrypto.so.3      [.] cipher_generic_block_update                                                                                                                         
   7.02%  openssl  libcrypto.so.3      [.] aesni_set_encrypt_key                                                                                                                               
   5.85%  openssl  libcrypto.so.3      [.] EVP_EncryptUpdate                                                                                                                                   
   5.13%  openssl  libpthread-2.17.so  [.] pthread_rwlock_unlock                                                                                                                               
   4.20%  openssl  libcrypto.so.3      [.] OPENSSL_cleanse                                                                                                                                     
   3.63%  openssl  libpthread-2.17.so  [.] pthread_rwlock_wrlock 

post patch
   8.97%  openssl  libc-2.17.so        [.] _int_malloc                                                                                                                                         
   8.50%  openssl  libcrypto.so.3      [.] aesni_set_encrypt_key                                                                                                                               
   5.95%  openssl  libc-2.17.so        [.] __strcmp_sse42                                                                                                                                      
   5.75%  openssl  libpthread-2.17.so  [.] pthread_rwlock_unlock                                                                                                                               
   5.35%  openssl  libpthread-2.17.so  [.] pthread_rwlock_wrlock                                                                                                                               
   5.17%  openssl  libcrypto.so.3      [.] OPENSSL_cleanse                                                                                                                                     
   5.14%  openssl  libcrypto.so.3      [.] ctr_update                                                                                                                                          
   5.13%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_key_length                                                                                                                           
   3.56%  openssl  libcrypto.so.3      [.] _aesni_encrypt3                                                                                                                                     
   3.09%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_reset                                                                                                                                
   3.05%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt                                                                                                                                   
   2.99%  openssl  libcrypto.so.3      [.] OSSL_PARAM_locate                                                                                                                                   
   2.82%  openssl  libcrypto.so.3      [.] cipher_generic_block_update
```

paulidale commented:
The `x86_64` timings are a bit worrying.

I'm guessing but I'd expect small generations to be common.

p-steuer commented:
> I'm guessing but I'd expect small generations to be common.

yes, that also bothers me. id like to be strictly better with this PR in all cases. i ll see what i can do

kroeckx commented:
#11102 is probably also going to help, I guess I should merge that.

I have some other ideas, but didn't find the time for them yet.

kroeckx commented:
Can you explain where the malloc and strcmp come from?

kroeckx commented:
I see a lot of calls to drbg_ctr_uninstantiate()?

mspncp commented:
@p-steuer which performance tool are you using to obtain the function-level statistics?

p-steuer commented:
linux perf tools: https://perf.wiki.kernel.org/index.php/Main_Page

For example:
```
perf record <some command>
perf report # shows function level statistics
```

p-steuer commented:
@kroeckx here are the expanded call-graphs for malloc and strcmp:
```
+   86.92%     0.00%  openssl  openssl             [.] _start                                                                                                                                 ◆
+   86.92%     0.00%  openssl  libc-2.17.so        [.] __libc_start_main                                                                                                                      ▒
+   86.92%     0.00%  openssl  openssl             [.] main                                                                                                                                   ▒
+   86.91%     0.02%  openssl  openssl             [.] speed_main                                                                                                                             ▒
+   86.90%     0.00%  openssl  openssl             [.] do_cmd                                                                                                                                 ▒
+   82.14%     0.17%  openssl  libcrypto.so.3      [.] RAND_DRBG_bytes                                                                                                                        ▒
+   81.34%     0.37%  openssl  libcrypto.so.3      [.] RAND_DRBG_generate                                                                                                                     ▒
+   81.01%     0.45%  openssl  libcrypto.so.3      [.] drbg_ctr_generate                                                                                                                      ▒
+   76.70%     4.97%  openssl  libcrypto.so.3      [.] ctr_update                                                                                                                             ▒
+   54.62%     2.51%  openssl  libcrypto.so.3      [.] EVP_CipherInit_ex                                                                                                                      ▒
+   14.99%     0.53%  openssl  libcrypto.so.3      [.] CRYPTO_zalloc                                                                                                                          ▒
+   13.22%     4.25%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_key_length                                                                                                              ▒
+   11.83%     0.00%  openssl  libc-2.17.so        [.] __GI___libc_malloc (inlined)                                                                                                           ▒
+   11.43%     3.00%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_reset                                                                                                                   ▒
+   11.09%     1.91%  openssl  libcrypto.so.3      [.] EVP_EncryptUpdate                                                                                                                      ▒
-   10.72%    10.72%  openssl  libc-2.17.so        [.] _int_malloc                                                                                                                            ▒
   - 10.16% _start                                                                                                                                                                            ▒
        __libc_start_main                                                                                                                                                                     ▒
      - main                                                                                                                                                                                  ▒
         - 10.16% do_cmd                                                                                                                                                                      ▒
              speed_main                                                                                                                                                                      ▒
              RAND_DRBG_bytes                                                                                                                                                                 ▒
              RAND_DRBG_generate                                                                                                                                                              ▒
              drbg_ctr_generate                                                                                                                                                               ▒
              ctr_update                                                                                                                                                                      ▒
            - EVP_CipherInit_ex                                                                                                                                                               ▒
               - 6.33% aes_256_ecb_newctx                                                                                                                                                     ▒
                    CRYPTO_zalloc                                                                                                                                                             ▒
                    __GI___libc_malloc (inlined)                                                                                                                                              ▒
                    _int_malloc                                                                                                                                                               ▒
               - 3.83% aes_256_ctr_newctx                                                                                                                                                     ▒
                    CRYPTO_zalloc                                                                                                                                                             ▒
                    __GI___libc_malloc (inlined)                                                                                                                                              ▒
                    _int_malloc                                                                                                                                                               ▒
     0.56% _int_malloc                                                                                                                                                                        ▒
+   10.66%     1.34%  openssl  libcrypto.so.3      [.] cipher_generic_get_ctx_params                                                                                                          ▒
+   10.51%     2.81%  openssl  libcrypto.so.3      [.] OPENSSL_init_crypto                                                                                                                    ▒
+   10.39%     0.47%  openssl  libcrypto.so.3      [.] engine_table_select_int                                                                                                                ▒
+    9.90%     2.90%  openssl  libcrypto.so.3      [.] OSSL_PARAM_locate                                                                                                                      ▒
+    9.57%     0.35%  openssl  libcrypto.so.3      [.] aes_256_ecb_newctx                                                                                                                     ▒
+    8.99%     0.55%  openssl  libcrypto.so.3      [.] cipher_hw_aesni_initkey                                                                                                                ▒
+    8.81%     2.61%  openssl  libcrypto.so.3      [.] cipher_generic_block_update                                                                                                            ▒
+    8.50%     8.50%  openssl  libcrypto.so.3      [.] aesni_set_encrypt_key                                                                                                                  ▒
+    8.50%     0.00%  openssl  libcrypto.so.3      [.] __aesni_set_encrypt_key (inlined)                                                                                                      ▒
+    6.71%     0.79%  openssl  libcrypto.so.3      [.] aes_256_ctr_newctx                                                                                                                     ▒
-    6.55%     6.55%  openssl  libc-2.17.so        [.] __strcmp_sse42      
     6.54% _start                                                                                                                                                                             ▒
        __libc_start_main                                                                                                                                                                     ▒
        main                                                                                                                                                                                  ▒
        do_cmd                                                                                                                                                                                ▒
        speed_main                                                                                                                                                                            ▒
        RAND_DRBG_bytes                                                                                                                                                                       ▒
        RAND_DRBG_generate                                                                                                                                                                    ▒
      - drbg_ctr_generate                                                                                                                                                                     ▒
         - 5.60% ctr_update                                                                                                                                                                   ◆
              EVP_CipherInit_ex                                                                                                                                                               ▒
              EVP_CIPHER_CTX_key_length                                                                                                                                                       ▒
              cipher_generic_get_ctx_params                                                                                                                                                   ▒
              OSSL_PARAM_locate                                                                                                                                                               ▒
              __strcmp_sse42                                                                                                                                                                  ▒
         - 0.94% EVP_CipherInit_ex                                                                                                                                                            ▒
              EVP_CIPHER_CTX_iv_length                                                                                                                                                        ▒
              cipher_generic_get_ctx_params                                                                                                                                                   ▒
              OSSL_PARAM_locate                                                                                                                                                               ▒
              __strcmp_sse42                                                                                                                                                                  ▒
+    5.94%     0.57%  openssl  libcrypto.so.3      [.] cipher_hw_aesni_ecb                                                                                                                    ▒
+    5.87%     0.53%  openssl  libcrypto.so.3      [.] CRYPTO_clear_free                                                                                                                      ▒
+    5.76%     0.21%  openssl  libcrypto.so.3      [.] CRYPTO_THREAD_unlock                                                                                                                   ▒
+    5.58%     2.39%  openssl  libcrypto.so.3      [.] aesni_ecb_encrypt                                                                                                                      ▒
+    5.55%     5.55%  openssl  libpthread-2.17.so  [.] pthread_rwlock_unlock                                                                                                                  ▒
+    5.55%     0.00%  openssl  libpthread-2.17.so  [.] __GI___pthread_rwlock_unlock (inlined)                                                                                                 ▒
+    5.54%     5.54%  openssl  libcrypto.so.3      [.] OPENSSL_cleanse                                                                                                                        ▒
+    4.83%     0.08%  openssl  libcrypto.so.3      [.] RAND_bytes_ex                                                                                                                          ▒
+    4.58%     0.54%  openssl  libcrypto.so.3      [.] CRYPTO_THREAD_write_lock                                                                                                               ▒
+    4.05%     0.00%  openssl  libpthread-2.17.so  [.] __GI___pthread_rwlock_wrlock (inlined)                                                                                                 ▒
+    4.04%     4.04%  openssl  libpthread-2.17.so  [.] pthread_rwlock_wrlock                                                                                                                  ▒
+    3.59%     1.50%  openssl  libcrypto.so.3      [.] ctr_BCC_update.part.2                                                                                                                  ▒
+    3.34%     0.10%  openssl  libcrypto.so.3      [.] OPENSSL_CTX_get0_public_drbg                                                                                                           ▒
+    3.26%     0.15%  openssl  libcrypto.so.3      [.] openssl_ctx_get_data                                                                                                                   ▒
+    3.14%     3.14%  openssl  libcrypto.so.3      [.] _aesni_encrypt3                                                                                                                        ▒
+    2.48%     2.48%  openssl  libc-2.17.so        [.] _int_free                                                                                                                              ▒
+    2.47%     2.47%  openssl  libc-2.17.so        [.] __memset_sse2                                                                                                                          ▒
+    2.18%     0.61%  openssl  libcrypto.so.3      [.] EVP_CIPHER_CTX_iv_length                
```

kroeckx commented:
So I tried 'valgrind --tool=callgrind ./apps/openssl speed -bytes 16 rand',
and it showed for 14147 calls to RAND_bytes():
%self called function
8.90  77813  _int_malloc
7.39  85014  OPENSSL_cleanse
6.90  56614  ctr_BCC_block.isra.0
6.27  70760  aesni_set_encrypt_key
4.10  424570 OSSL_PARAM_locate
3.88  127372 cipher_generic_block_update
3.41  427106 __strcmp_sse2_unaligned
3.04  77812  _int_free
3.03  84910  EVP_CipherInit_ex
3.02  141522 EVP_EncryptUpdate
2.89  28302  ctr_XOR.part.0
2.61  70897  OPENSSL_init_crypto
2.59  84916  _aesni_encrypt3
2.40  127372 aesni_ecb_encrypt
1.95  101610 _memset_sse2_unaligned
1.81  28302  ctr_update
[...]

p-steuer commented:
The problem was that I used EVP_CipherInit_ex with a non-null cipher argument even when updating the key/iv. This lead to the cipher objects being erased and re-allocated on every call.

Now that its fixed, the theoretical perf. gain shows up also at smaller sizes:
```
x86_64
type          16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
rand pre      11174.77k    41708.71k   136128.51k   321467.73k   536630.61k   559355.22k
rand post     16731.10k    64872.32k   237013.85k   794792.96k  2609661.60k  3103227.90k
```
(#11102 was already included for above measurement. It helps both pre- and post- perf perf.)

p-steuer commented:
red cross due to timeouts

p-steuer commented:
So i think this one is ready to go. Please re-review.

t8m commented on crypto/rand/drbg_ctr.c:
```diff
@@ -84,12 +85,16 @@ __owur static int ctr_BCC_block(RAND_DRBG_CTR *ctr, unsigned char *out,
-        return 0;
-    return 1;
+    uint8_t in_tmp[48];
+    uint8_t num_of_blk = 2;
```
I would refrain from using uint8_t here. The ctr_BCC_block parameters are unsigned char * and int so I would use them.

t8m commented on crypto/rand/drbg_ctr.c:
```diff
@@ -98,19 +103,14 @@ __owur static int ctr_BCC_blocks(RAND_DRBG_CTR *ctr, const unsigned char *in)
  */
 __owur static int ctr_BCC_init(RAND_DRBG_CTR *ctr)
 {
+    uint8_t bltmp[48] = {0};
```
The same here.

t8m commented on crypto/rand/drbg_ctr.c:
```diff
@@ -231,31 +231,25 @@ __owur static int ctr_update(RAND_DRBG *drbg,
 {
     RAND_DRBG_CTR *ctr = &drbg->data.ctr;
     int outlen = AES_BLOCK_SIZE;
+    uint8_t V_tmp[48], out[48];
```
And here.

t8m commented:
One of the Travis jobs timed out in test_rand_extra so probably the number of data generated in this test should be reconsidered.

p-steuer commented:
I adressed the uint8_t -> unsigned char thing, disabled 2/3 test vectors in test_rand_extra to decrease runtime and rebased.

p-steuer commented:
this time travis red cross looks unrelated to this PR

p-steuer commented:
One build still timed out, so I included the test case in rand_test but commented it out.

p-steuer commented:
hmm the build time outs at fuzz test now, dont know why.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

p-steuer commented:
merged.

master: 28bdbe1aaa474ae8cd83e520d02e463e46ce89d9 4d6d787c4937706676355ce241c6e538475295d2
